src/OrleansAzureUtils/MultiClusterNetwork/AzureTableBasedGossipChannel.cs

```diff
@@ -0,0 +1,202 @@
+        {
+            Name = "AzureTableBasedGossipChannel-" + ++sequenceNumber;
+            logger = TraceLogger.GetLogger(Name, TraceLogger.LoggerType.Runtime);
+
+            logger.Info("Initializing Gossip Channel for ServiceId={0} using connection: {1}, SeverityLevel={2}",
+                globalconfig.GlobalServiceId, ConfigUtilities.RedactConnectionStringInfo(connectionstring), logger.SeverityLevel);
+
+            tableManager = await GossipTableInstanceManager.GetManager(globalconfig.GlobalServiceId, connectionstring, logger);
+        }
+
+        // used by unit tests
+        public Task DeleteAllEntries()
+        {
+            logger.Info("DeleteAllEntries");
+            return tableManager.DeleteTableEntries();
+        }
+
+      
+        private static void UpdateDictionaryRightValue<TKey, TLeft, TRight>(Dictionary<TKey, KeyValuePair<TLeft, TRight>> dict, TKey key, TRight rightValue)
+        {
+            if (dict.ContainsKey(key))
+                dict[key] = new KeyValuePair<TLeft, TRight>(dict[key].Key, rightValue);
+            else
+                dict.Add(key, new KeyValuePair<TLeft, TRight>(default(TLeft), rightValue));
+        }
+
+        // IGossipChannel
+        public async Task Push(MultiClusterData data)
+        {
+            logger.Verbose("-Push data:{0}", data);
+
+            // this is typically called with just one, or very few, items in data
+            // thus, it is sufficiently performant to read rows individually
+
+            var retrievalTasks = new List<Task<GossipTableEntry>>();
+            if (data.Configuration != null)
+            {
+                retrievalTasks.Add(tableManager.ReadConfigurationEntryAsync());
+            }
+
+            foreach (var gateway in data.Gateways.Values)
+            {
+                retrievalTasks.Add(tableManager.ReadGatewayEntryAsync(gateway));
+            }
+
+            await Task.WhenAll(retrievalTasks);
+
+            var entriesFromStorage = retrievalTasks.Select(t => t.Result).Where(entry => entry != null);
+            await DiffAndWriteBack(data, entriesFromStorage); 
+        }
+
+        // IGossipChannel
+        public async Task<MultiClusterData> PushAndPull(MultiClusterData pushed)
+        {
+            logger.Verbose("-PushAndPull pushed:{0}", pushed);
+
+            try
+            {
+                var entriesFromStorage = await tableManager.FindAllGossipTableEntries();
+                var delta = await DiffAndWriteBack(pushed, entriesFromStorage);
+
+                logger.Verbose("-PushAndPull pulled delta:{0}", delta);
+
+                return delta;
+            }
+            catch (Exception e)
+            {
+                logger.Info("-PushAndPull encountered exception {0}", e);
+
+                throw e;
+            }
+        }
+
+        internal async Task<MultiClusterData> DiffAndWriteBack(MultiClusterData dataToPush, IEnumerable<GossipTableEntry> entriesFromStorage)
```

Comment:
I have separated the config/gateway parts, so it is a bit better.

Still, diffing and writeback are in the same method because the case distinction in the diffing is needed in order to know whether/what to write back.